[Flang][MLIR][OpenMP] Create a deferred declare target marking process for Bridge.cpp

agozillon opened the pull request:
This patch seeks to create a process that happens on module finalization for OpenMP, in which a list of operations that had declare target directives applied to them and were not generated at the time of processing the original declare target directive are re-checked to apply the appropriate declare target semantics.

This works by maintaining a vector of declare target related data inside of the FIR converter, in this case the symbol and the two relevant unsigned integers representing the enumerators. This vector is added to via a new function called from Bridge.cpp, insertDeferredDeclareTargets, which happens prior to the processing of the directive (similarly to getDeclareTargetFunctionDevice currently for requires), it effectively checks if the Operation the declare target directive is applied to currently exists, if it doesn't it appends to the vector. This is a seperate function to the processing of the declare target via the overloaded genOMP as we unfortunately do not have access to the list without passing it through every call, as the AbstractConverter we pass will not allow access to it (I've seen no other cases of casting it to a FirConverter, so I opted to not do that).

The list is then processed at the end of the module in the finalizeOpenMPLowering function in Bridge by calling a new function markDelayedDeclareTargetFunctions which marks the latently generated operations. In certain cases, some still will not be generated, e.g. if an interface is defined, marked as declare target, but has no definition or usage in the module then it will not be emitted to the module, so due to these cases we must silently ignore when an operation has not been found via it's symbol.

The main use-case for this (although, I imagine there is others) is for processing interfaces that have been declared in a module with a declare target directive but do not have their implementation defined in the same module. For example, inside of a seperate C++ module that will be linked in. In cases where the interface is called inside of a target region it'll be marked as used on device appropriately (although, realistically a user should explicitly mark it to match the corresponding definition), however, in cases where it's used in a non-clear manner through something like a function pointer passed to an external call we require this explicit marking, which this patch adds support for (currently will cause the compiler to crash).

An alternative to this approach may be to do it in the implicit marking pass, by searching for function pointers/addresses used in target regions and marking the FuncOp's appropriately. But this doesn't seem like the right approach, we would be ignoring the users explicit markings at that point if they exist (we'd have no way of saving the data and transferring it to the pass trivially) in lieu of our own assumptions, and this would also not support other possible operation types that may need deferred marking.

agozillon committed:
[Flang][MLIR][OpenMP] Create a deferred declare target marking process for Bridge.cpp

This patch seeks to create a process that happens apon module
finalisation for OpenMP, in which a list of operations that had
declare target directives applied to them and were not generated
at the time of processing the original declare target directive are
re-checked to apply the appropriate declare target semantics.

This works by maintaining a vector of declare target related data
inside of the FIR converter, in this case the symbol and the two
relevant unsigned integers representing the enumerators. This
vector is added to via a new function called from Bridge.cpp,
insertDeferredDeclareTargets, which happens prior to the
processing of the directive (similarly to getDeclareTargetFunctionDevice
currently for requires), it effectively checks if the Operation the
declare target directive is applied to currently exists, if it doesn't it
appends to the vector. This is a seperate function to the processing
of the declare target via the overloaded genOMP as we unfortunately
do not have access to the list without passing it through every call,
as the AbstractConverter we pass will not allow access to it (I've seen
no other cases of casting it to a FirConverter, so I opted to not do that).

The list is then processed at the end of the module in the
finalizeOpenMPLowering function in Bridge by calling a
new function markDelayedDeclareTargetFunctions which
marks the latently generated operations. In certain cases,
some still will not be generated, e.g. if an interface is defined,
marked as declare target, but has no definition or usage in the
module then it will not be emitted to the module, so due to
these cases we must silently ignore when an operation has
not been found via it's symbol.

The main use-case for this (although, I imagine there is others) is
for processing interfaces that have been declared in a module with
a declare target directive but do not have their implementation
defined in the same module. For example, inside of a seperate
C++ module that will be linked in. In cases where the interface
is called inside of a target region it'll be marked as used on
device appropriately (although, realistically a user should explicitly
mark it to match the corresponding definition), however, in
cases where it's used in a non-clear manner through
something like a function pointer passed to an external call
we require this explicit marking, which this patch adds
support for (currently will cause the compiler to crash).

An alternative to this approach may be to do it in the implicit marking pass,
by searching for function pointers/addresses used in target regions and
marking the FuncOp's appropriately. But this doesn't seem like the right
approach, we would be ignoring the users explicit markings at that point
if they exist (we'd have no way of saving the data and transferring it to the
pass trivially) in lieu of our own assumptions, and this would also not support
other possible operation types that may need deferred marking.

agozillon committed:
[NFC] Remove some extra braces clang-format may or may not have added

ergawy reviewed (commented):
Still working on understanding the details of changes introduced in the PR. But for now just a few quick remarks from an initial look at the first few files.

agozillon commented on flang/include/flang/Lower/OpenMP.h:
```diff
@@ -84,6 +87,19 @@ bool isOpenMPTargetConstruct(const parser::OpenMPConstruct &);
 bool isOpenMPDeviceDeclareTarget(Fortran::lower::AbstractConverter &,
                                  Fortran::lower::pft::Evaluation &,
                                  const parser::OpenMPDeclarativeConstruct &);
+void gatherDeferredDeclareTargets(
+    Fortran::lower::AbstractConverter &, Fortran::lower::pft::Evaluation &,
+    const parser::OpenMPDeclarativeConstruct &,
+    llvm::SmallVectorImpl<std::tuple<
+        uint32_t /*mlir::omp::DeclareTargetCaptureClause*/, uint32_t,
+        /*mlir::omp::DeclareTargetDeviceType*/ Fortran::semantics::Symbol>> &);
+bool markDelayedDeclareTargetFunctions(
+    mlir::Operation *,
+    llvm::SmallVectorImpl<
+        std::tuple<uint32_t /*mlir::omp::DeclareTargetCaptureClause*/,
+                   uint32_t, /*mlir::omp::DeclareTargetDeviceType*/
+                   Fortran::semantics::Symbol>> &,
+    AbstractConverter &);
```
Not sure what you mean by this unfortunately, if you could elaborate more that'd be appreciated!

agozillon commented on flang/include/flang/Lower/OpenMP.h:
```diff
@@ -84,6 +87,19 @@ bool isOpenMPTargetConstruct(const parser::OpenMPConstruct &);
+    llvm::SmallVectorImpl<std::tuple<
+        uint32_t /*mlir::omp::DeclareTargetCaptureClause*/, uint32_t,
+        /*mlir::omp::DeclareTargetDeviceType*/ Fortran::semantics::Symbol>> &);
+bool markDelayedDeclareTargetFunctions(
```
Good idea, happy to do a type alias, however, I'd lean towards keeping it a tuple as that seems to be the standard in most cases, but I'm not overly bothered either way, so happy to change it if you feel strongly about it!

agozillon commented:
Small ping for some reviewer attention next week if any time can be spared

agozillon commented:
Another small ping for reviewer attention if at all possible before I go ahead and address the existing comments. 

I am primarily interested in if this seems like a reasonable approach to the problem from @kiranchandramohan or @jeanPerier and if not possible suggestions towards an alternative!

clementval commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -5015,6 +5025,17 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+                 uint32_t, /*mlir::omp::DeclareTargetDeviceType*/
+                 Fortran::semantics::Symbol>,
+      2>
+      deferredDeclareTarget;
```
Can the name be more explicit with an omp prefix of smth?

agozillon replied:
It can indeed, I'll also try and make the previously suggested typedef more OpenMP explicit as well.

agozillon commented:
Updated the PR based on reviewer comments [(commit)](https://github.com/llvm/llvm-project/pull/78502/commits/3feda65753323823e226f9eda0ec25868f8d93e1#diff-0f70165af00b05acb5abeb11198079b793a20acf8f069edd2dbe7b34ac67b328). 

I rebased as it's been a little while since the PR has been updated fully, a side affect of that is the need to force push, so I apologies to reviewers if that makes things difficult when reviewing the recent changes. 

The recent changes encompass some minor changes: 

- Moved to using a type-alias for the tuple
- Made names a little more consistent and OpenMP specific across the board

agozillon commented:
Small ping to see if we can move this PR forward one way or another! Thank you very much ahead of time for any of your precious time you can give to the review! :-)

kiranchandramohan commented:
@agozillon This looks OK, but the code that handles `declare target` has changed quite a bit after the initial patch. Since it is split across different places (lowering, pass, translation to LLVM), it would be good to document this.

agozillon commented:
Thank you Kiran, I'll perform a rebase and add a document of the overall process as I understand it now, and @skatrak who has also worked on it can verify if I am incorrect in any places when it's up and he has time!

agozillon committed:
[NFC] Update based on review comments

> Moved to using a type-alias for the tuple
> Made names a little more consistent and OpenMP specific

agozillon commented:
Recent commit performed a rebase (nothing too interesting change wise, just re-aligning certain function calls with some recent changes).

It also added new documentation on declare target handling, I would appreciate if you could have a little look at the parts that you've worked on @skatrak to make sure I've captured the essence of it!

agozillon committed:
[NFC] Add documentation for declare target as requested

agozillon committed:
Fix document build bugs

agozillon commented:
Most recent PR (hopefully) fixes the documentation build bug I introduced:

* Actually add the document to the index file, which I forgot to do
* change the MLIR code block into a blank code block in the new MD as it's unsupported at the moment it seems

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,124 @@
+
+In the case where we encounter a `FuncOp` our processing is fairly simple, if we're processing the device code, we will finish up our removal of `host` marked functions, anything we could not remove previously we now remove, e.g. if it had a `target` directive in it (which we need to keep a hold of to this point, to actually outline the `target` kernel for device). This hopefully leaves us with o ...
+
+For `GlobalOp`'s, the processing is a little more complex, we currently leverage two OMPIRBuilder functions which we have inherited from Clang and moved to the `OMPIRBuilder` to share across the two compiler frontends `registerTargetGlobalVariable` and `getAddrOfDeclareTargetVar`. These two functions are actually recursive and invoke each other depending on the clauses and options provided to the ...
```
```suggestion
For `GlobalOp`'s, the processing is a little more complex. We currently leverage the `registerTargetGlobalVariable` and `getAddrOfDeclareTargetVar` `OMPIRBuilder` functions shared with Clang. These two functions invoke each other depending on the clauses and options provided to the `OMPIRBuilder` (in particular, unified shared memory). Their main purposes are the generation of a new global device pointer with a "ref_" prefix on the device and enqueuing metadata generation by the `OMPIRBuilder` to be produced at module finalization time. This is done for both host and device and it links the newly generated device global pointer and the host pointer together across the two modules. 
```

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,124 @@
+
+For `GlobalOp`'s, the processing is a little more complex, we currently leverage two OMPIRBuilder functions which we have inherited from Clang and moved to the `OMPIRBuilder` to share across the two compiler frontends `registerTargetGlobalVariable` and `getAddrOfDeclareTargetVar`. These two functions are actually recursive and invoke each other depending on the clauses and options provided to the ...
+
+Two things of note about the `GlobalOp` processing, the first is that similarly to other metadata (e.g. for `TargetOp`) that is shared across both host and device modules, the device needs access to the previously generated host IR file, which is done through another `attribute` applied to the `ModuleOp` by the compiler frontend. The file is loaded in and consumed by the `OMPIRBuilder` to populat ...
```
```suggestion
Similarly to other metadata (e.g. for `TargetOp`) that is shared across both host and device modules, processing of `GlobalOp`'s in the device needs access to the previously generated host IR file, which is done through another `attribute` applied to the `ModuleOp` by the compiler frontend. The file is loaded in and consumed by the `OMPIRBuilder` to populate it's `OffloadInfoManager` data structures, keeping host and device appropriately synchronised.
```

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,124 @@
+    * Descriptor based types, e.g. pointers/allocatables.
+    * Derived types.
+    * Members of derived types (use-case needs legality checking with OpenMP specification).
+* Marking global variables with `declare target`'s `to` clause, a lot of the lowering should exist, but it needs further testing and likely some further changes to fully function.
```
```suggestion
* Marking global variables with `declare target`'s `to` clause. A lot of the lowering should exist, but it needs further testing and likely some further changes to fully function.
```

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,124 @@
+   and the `C` function in the other module must be marked `declare target`, with the same type of 
+   marking as indicated by the specification.
+* Marking global variables with `declare target`'s `link` clause and mapping the data to the device data 
+   environment utilising `declare target` (may not work for all types yet, but for scalars and arrays
```
```suggestion
   environment utilising `declare target`. This may not work for all types yet, but for scalars and arrays
```

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,124 @@
+   marking as indicated by the specification.
+* Marking global variables with `declare target`'s `link` clause and mapping the data to the device data 
+   environment utilising `declare target` (may not work for all types yet, but for scalars and arrays
+   of scalars, it should).
```
```suggestion
   of scalars, it should.
```

skatrak commented on flang/docs/index.md:
```diff
@@ -69,6 +69,7 @@ on how to get in touch with us and to learn more about the current status.
    OpenACC-descriptor-management.md
    OpenMP-4.5-grammar.md
    OpenMP-descriptor-management
+   OpenMP-declare-target
```
Nit: Move one line up for alphabetical order.

skatrak reviewed (commented):
Thank you Andrew for this work. I think the overall approach makes sense, my comments are just over a few details.

I added quite a few suggestions to the documentation, to try to make it a bit easier to follow and perhaps clarify a couple of points. Feel free to ignore any suggestions there that you think don't improve the text.

agozillon commented on flang/include/flang/Lower/OpenMP.h:
```diff
@@ -49,6 +52,11 @@ struct Evaluation;
 } // namespace pft
 
+using OMPDeferredDeclTarInfo =
+    std::tuple<uint32_t /*mlir::omp::DeclareTargetCaptureClause*/,
+               uint32_t, /*mlir::omp::DeclareTargetDeviceType*/
+               Fortran::semantics::Symbol>;
```
Can try to do, it was intended to mimic the existing DeclareTargetCapturePair we have already, I can maybe use the dialect enums, it depends how bothered we are about including the dialect header or forward declaring them (i recall forward declaring being a little bit awkward, but it's been a while)

agozillon commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -5057,6 +5068,13 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+  /// processed at the time of lowering the declare target construct, such
+  /// as certain cases where interfaces are declared but not defined within
+  /// a module.
+  llvm::SmallVector<Fortran::lower::OMPDeferredDeclTarInfo, 2>
```
Will do it as part of this PR as it's a small change and makes swapping the contents of this patch to a ref a little simpler (we get the symbol from the pair, which isn't a problem when taking a copy, but is if we take a ref currently as I believe the copy in the pair falls out of scope before we process the deferred values, swapping both to ref means we're pointing at the original which persists to and past when we require it)

agozillon commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -3959,42 +4015,16 @@ static void genOMP(Fortran::lower::AbstractConverter &converter,
-    // appear in the symbol table as they are represented as allocas
+
+    // Some symbols are deferred until later in the module, these are handled
+    // apon finalization of the module for OpenMP inside of Bridge, so we simply
```
I think you've corrected me on this exact same spelling mistake a few times now... thank you very much :D might be time to integrate a spell checker into my IDE... especially for the documentation component

agozillon committed:
Update documentation and implementation based on reviewer feedback

agozillon commented:
Thank you Sergio for the feedback! I believe I've addressed it all in the recent commit, which should: 
> Fix up the documentation with recommended changes
> Fix alphabetical ordering of the index
> Change to a struct from a tuple
> Add a TODO message

agozillon committed:
required DeclareTargetCapturePair change lost in rebase

agozillon commented:
Performed a rebase on recent refactoring to the OpenMP lowering, nothing too drastic required in this PR it is mostly the same!

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,259 @@
+
+Whenever we invoke `genFIR` on an `OpenMPDeclarativeConstruct` from the 
+lowering bridge, we are also invoking another function called 
+`gatherOpenMPDeferredDeclareTargets` which gathers information relevant to the
```
```suggestion
`gatherOpenMPDeferredDeclareTargets`, which gathers information relevant to the
```

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,259 @@
+first phase. The main notable case this occurs currently is when a 
+Fortran function interface has been marked. This is 
+done via the function 
+`markOpenMPDeferredDeclareTargetFunctions` which is called from the lowering
```
```suggestion
`markOpenMPDeferredDeclareTargetFunctions`, which is called from the lowering
```

skatrak commented on flang/lib/Lower/Bridge.cpp:
```diff
@@ -5078,6 +5088,13 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+  /// processed at the time of lowering the declare target construct, such
+  /// as certain cases where interfaces are declared but not defined within
+  /// a module.
+  llvm::SmallVector<Fortran::lower::OMPDeferredDeclareTargetInfo, 2>
```
I think we should be able to use the default now.
```suggestion
  llvm::SmallVector<Fortran::lower::OMPDeferredDeclareTargetInfo>
```

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,259 @@
+applied to any operation in any dialect. This helps to simplify the 
+utilisation of it, instead of replacing or modifying existing global or 
+function `operations` in a dialect it applies to it as extra metadata that 
+the lowering can use in different ways as is necesary. 
```
```suggestion
the lowering can use in different ways as is necessary. 
```

skatrak commented on flang/docs/OpenMP-declare-target.md:
```diff
@@ -0,0 +1,259 @@
+remove or delete the old symbol, this is due to the fact that the same symbol
+can be utilised across multiple target regions, if we remove it, we risk 
+breaking lowerings of target regions that will be processed at a later time. 
+To appropriately delete these no longer necesary symbols we would need a 
```
```suggestion
To appropriately delete these no longer necessary symbols we would need a 
```

skatrak reviewed (commented):
Thank you for addressing my comments Andrew, I've got a few nits left after a second look, but they should be trivial and this looks nearly ready to me.

I marked as resolved my prior comments that have been addressed, but there are a couple that I left open because they don't seem to have been incorporated yet. Let me know if this is just because you don't think these documentation suggestions improve readability.

agozillon committed:
Address reviewer comments

skatrak reviewed (approved):
Thank you Andrew for addressing my comments. This LGTM, but leave some time before merging in case @ergawy, @kiranchandramohan or someone else has any further suggestions.

agozillon commented:
Thank you very much Sergio for the detailed review, I appreciate it greatly :-) as well as @ergawy's, I think the PR is in much better shape now. 

More than happy to wait a little bit, it would be nice to get one more approval on the PR before landing!

kiranchandramohan reviewed (approved):
LG

agozillon commented:
Thank you very much @kiranchandramohan ! I'll land the PR just now then as it's passed the CI and has no conflicts.

As for the presentation I'd be happy to do it at some point, although, I am admittedly not the best presenter! Myself and @skatrak  can likely discuss who would prefer to do it in the future offline.
